### PR TITLE
WMDP-193: Generate eligibility screener CSV file

### PR DESCRIPTION
**Pull request by chouinar**
## Ticket

https://wicmtdp.atlassian.net/browse/WMDP-193

Note this is built ontop of the changes in PR https://github.com/navapbc/wic-mt-demo-project-mock-api/pull/13

## Changes
Added a script that creates a CSV of eligibility records.

Added a few minor utilities for setting up scripts.

## Context for reviewers
Kept the naming of the columns the same as the prototype, but setup to be easy to adjust.

A timestamp column is used to determine if the record has been added to a report yet. When we finish generating a CSV, this timestamp gets updated.

## Testing

DB before running the script:
Screen Shot 2022-09-06 at 4 50 50 PM

Run the script with `make create-eligibility-screener-csv` and you can see the timestamp updated in the DB
Screen Shot 2022-09-06 at 4 51 28 PM

The CSV is generated locally:
Screen Shot 2022-09-06 at 4 51 44 PM
and in excel:
Screen Shot 2022-09-06 at 4 52 24 PM

### CODE REVIEW

File: app/api/db/models/eligibility_models.py [hunk 1]
```diff
@@ -16,18 +16,17 @@ class EligibilityScreener(Base, TimestampMixin):
...
     has_prior_wic_enrollment = Column(Boolean, nullable=False)
     eligibility_programs: list[str] = Column(MutableList.as_mutable(ARRAY(Text)), nullable=True)
     household_size = Column(Integer, nullable=True)
-    zip_code = Column(Text, nullable=False)
+    zip_code: str = Column(Text, nullable=False)
     applicant_notes = Column(Text, nullable=True)
 
-    # TODO - when we build out the CSV generation we likely want some sort
-    # of column that indicates it was put into the CSV.
+    added_to_eligibility_screener_at = Column(TIMESTAMP(timezone=True), nullable=True)
```
Comment (rocketnova) [inline app/api/db/models/eligibility_models.py]: Should we add typing for `has_prior_wic_enrollment`, `household_size`, `applicant_notes`, and `added_to_eligibility_screener_at`?

File: app/api/scripts/create_eligibility_screener_csv.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+def convert_eligibility_screener_records_for_csv(
+    records: list[EligibilityScreener],
+) -> list[EligiblityScreenerCsvRecord]:
+    logger.info("Converting eligibility screeners to CSV format")
+    out_records: list[EligiblityScreenerCsvRecord] = []
+    out_records.append(ELIGIBILITY_SCREENER_CSV_HEADER)
+
+    for record in records:
+        # TODO - do we want to convert this timestamp to a different timezone beside UTC?
```
Comment (rocketnova) [inline app/api/scripts/create_eligibility_screener_csv.py:96]: Oh great question. I think yes. We should set this to CT. Can we make this an env var?

File: app/api/scripts/create_eligibility_screener_csv.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    return out_records
+
+
+def generate_csv_file(records: list[EligiblityScreenerCsvRecord], output_file_path: str) -> None:
+    logger.info("Generating eligibility screener CSV at %s", output_file_path)
+    # TODO - when we want to also support S3, change this to
+    # use the smart_open library which will handle S3 + local files
+
+    # TODO - if no records, should we be fine just making a file
+    # with the headers or do we want to do something more?
```
Comment (rocketnova) [inline app/api/scripts/create_eligibility_screener_csv.py:126]: Good question. My inclination is to just output a CSV that is basically empty and just has headers. Ping @laurjpeterson thoughts?

File: app/api/scripts/create_eligibility_screener_csv.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            chosen_wic_clinic="TODO",
+            applicant_notes=blank_for_null(record.applicant_notes),
+            submitted_datetime=submitted_datetime,
+        )
+        out_records.append(out_record)
+
+    return out_records
+
+
+def generate_csv_file(records: list[EligiblityScreenerCsvRecord], output_file_path: str) -> None:
```
Comment (rocketnova) [inline app/api/scripts/create_eligibility_screener_csv.py:120]: For the platform work (so I don't mean necessarily for this PR), I would encourage you to break out csv utilities into a separate util file).

File: app/tests/api/scripts/test_create_eligibility_screener_csv.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+
+def validate_csv_records(db_records, csv_records, test_db_session):
+    assert len(csv_records) == len(db_records)
+
+    # Sort the two lists by first name and zip together for validation
+    csv_records.sort(key=lambda record: record["First Name"])
+    db_records.sort(key=lambda record: record.first_name)
+    for csv_record, db_record in zip(csv_records, db_records):
+        test_db_session.refresh(db_record)
```
Comment (rocketnova) [inline app/tests/api/scripts/test_create_eligibility_screener_csv.py:38]: Can you remind me what `refresh()` does again at this point?

File: app/tests/api/scripts/test_create_eligibility_screener_csv.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    )
+
+    create_eligibility_screener_csv(test_db_session, tmp_file_path)
+    csv_rows = read_csv_records(tmp_file_path)
+    validate_csv_records(db_records, csv_rows, test_db_session)
+
+    # If we run it again, just a blank file should be made
+    create_eligibility_screener_csv(test_db_session, tmp_file_path)
+    csv_rows = read_csv_records(tmp_file_path)
+    assert len(csv_rows) == 0
```
Comment (rocketnova) [inline app/tests/api/scripts/test_create_eligibility_screener_csv.py:107]: Why doesn't this return 1 row for the header?

**Review by rocketnova — COMMENTED** [review]
Looks great. Some questions inline.

File: app/api/db/models/eligibility_models.py (hunk 1 above)
Comment (chouinar) [reply app/api/db/models/eligibility_models.py]: mypy wasn't complaining about being unable to infer the type from sqlalchemy (mainly couldn't determine `nullable=True` means optional), but I'll add it anyways for consistency.

File: app/api/scripts/create_eligibility_screener_csv.py (hunk 2 above)
Comment (chouinar) [reply app/api/scripts/create_eligibility_screener_csv.py:96]: Yes, went ahead and added a utility for converting datetime objects between timezones while I was at it.

File: app/api/scripts/create_eligibility_screener_csv.py (hunk 4 above)
Comment (chouinar) [reply app/api/scripts/create_eligibility_screener_csv.py:120]: I agree. There are additional CSV utilities I think would be good to bring over.

File: app/tests/api/scripts/test_create_eligibility_screener_csv.py (hunk 5 above)
Comment (chouinar) [reply app/tests/api/scripts/test_create_eligibility_screener_csv.py:38]: It updates the object with whatever has been flushed to the DB. In this case, the refresh is to get the timestamp that was added to another object representing the same DB record.

File: app/tests/api/scripts/test_create_eligibility_screener_csv.py (hunk 6 above)
Comment (chouinar) [reply app/tests/api/scripts/test_create_eligibility_screener_csv.py:107]: The parser by default assumes the first row is the header, and subsequent rows are actual data. The function is returning the accumulated list of iterating over the dict_reader, which is empty.

Comment (rocketnova) [reply app/tests/api/scripts/test_create_eligibility_screener_csv.py:107]: Oh got it. It's the difference between `convert_eligibility_screener_records_for_csv()` and `csv.DictReader()`. Thanks!

**Review by rocketnova — APPROVED** [review]
LGTM!

File: app/api/scripts/create_eligibility_screener_csv.py (hunk 3 above)
Comment (lpciferri) [reply app/api/scripts/create_eligibility_screener_csv.py:126]: I agree! I think just the headers makes sense.

**Pull request merged**